ideas/000-discoverable-trusted-server-nodes/README.md

```diff
@@ -0,0 +1,73 @@
+---
+id: 000-discoverable-trusted-server-nodes
+title: Discoverable And Trusted Server Nodes
+status: Draft
+created: 2018-06-07
+category: core
+lead-contributor: adambabik
+contributors:
+exit-criteria: yes
+success-metrics: yes
+clear-roles: no
+future-iterations: yes
+roles-needed:
+    - Smart Contract Developer
+    - Go Developer
+    - Clojure Developer
+    - UI/UX Designer
+okrs:
+    - "[PO]: Objective: Messaging is reliable"
+    - "[P2]: KR: Cluster can handle 500 concurrent users"
+---
+
+## Summary
+The idea here is to provide a way to make Mail Server nodes discoverable and trusted. This should work seamlessly for the end user.
+
+## Product Overview
+Currently, Mail Servers (and other planned server nodes like Push Notification servers) need to be hardcoded in the app code. It's possible to provide your own Mail Server in the app settings but that's not a scalable solution and it does not solve all issues.
+
+The outcome of this swarm are fully decentralized and discoverable Mail Servers. The Status app will have a mean to verify that a discovered Mail Server is trusted.
+
+In the next iterations, we may work on a way to introduce external providers to submit their own Mail Servers into the cluster and allow end users to accept these resources as trusted.
+
+## Product Description
+The work in this swarm can be divided into three areas:
+- (Infra) making Mail servers discoverable,
+- (Smart Contract) writing a smart contract to keep track of trusted Mail Servers,
+- (Client) mechanism of verifying whether a discovered Mail Server is trusted.
+
+As mentioned in the Product Overview, the smart contract should allow registering custom providers. The client will have an option to decide which providers it trust. Status will be a default provider and by default the app will trust it.
```

Comment:
> This smart contract should allow, not only adding, but also removing trusted mail servers (in case they become blocked, corrupted or dead).

Yeah, all management functions must be implemented.

> i don't think its necessary to communicate about those untrusted mail servers

The thing is that anyone else can advertise anything in our network. We don't check if a node comes from our cluster or from outside in bootnodes.

Reply:
So, from the quote:
> 1. [ ] A client that collects discovered Mail Servers and verifies if they are trusted using the smart contract. Verified Mail Servers are added to the pool and a random one is selected to use.

We can .... a workflow like:
```
1.- client subscribes to a discovery channel
2.- client receives a broadcast from a mail server (MS1)
3.- client verifies the mail server (MS1) against the smart contract
4.- mail server (MS1) is verified, and added to client's mail server pool
5.- smart contract receives an order to untrust (MS1)
6.- mail server (MS1) is removed from trusted list on smart contract
```
At this point the client will have MS1 on its pool of trusted mail servers, however its not a trusted by the smart contract anymore.

Then we have 2 ways of proceed here:

1.- Query the smart contract on every `RequestHistoricMessages` call
2.- Have a process running on the client to verify its pool of mail servers are still trusted by the smart contract.

Reply:
It will work a bit different with Discovery V5.

1. Mail Server registers itself with a topic `mailserver`,
2. A client subscribes to Discovery V5 for peers registered with `mailserver` topic,
3. A client verifies that a found Mail Server is trusted,
3.1. If so, the Mail Server is added to the pool,
3.2. Otherwise, it's discarded.

Now, how do we handle the fact that a selected Mail Server was removed from the trusted list? I think we can use smart contract events (should be confirmed with someone from @status-im/clojure team). In this case, the client does not need to check the Mail Server before every `RequestHistoricMessages`. Only when starting for the first time or depending on the characteristic of sending smart contract events.

Reply:
@adambabik just confirming we can `watch` for events generated by the contract (new) or fetch old events. So checking once then `watching` as mentioned should work.